Implement pause-aware picking (3.2)

RandomShaper opened the pull request:
**NOTES:**
- The diff looks super big, but that's because I've moved part of the logic in `Viewport`'s `INTERNAL_PHYSICS_PROCESS` to a function. The logic changes in that section of code are some checks to let it work with either the classic or the new behavior.
- This PR is subject to discussion. The right thing to do may not be to have this as a project setting, but trying to find a behavior that involves the good features of the new one while keeping compatibility.
- Also, in my opinion, the new agreed behavior should become the default and only in 4.0. **UPDATE:** PR here: #45775

---

This adds a new project setting (`physics/common/enable_pause_aware_picking`). It's disabled by default.

When enabled, it changes the way 2D & 3D physics picking behaves in relation to pause:
- When pause is set, every collision object that is hovered or captured (3D only) is released from that condition, getting the relevant mouse-exit callback, unless its pause mode makes it immune to pause.
- During the pause, picking only considers collision objects immune to pause, sending input events and enter/exit callbacks to them as expected.
- When pause is left, nothing happens. This is a big difference with the classic behavior, which at this point would process all the input events that have been queued, applying them to the current state of the 2D/3D world (in other words, checking them against the current position of the objects instead of those at the time of the events).

Fixes #3703.

Zireael07 commented:
Any news on this? It's been laying about for 2 months and it fixes a fairly important issue...

Calinou commented:
@Zireael07 Akien is currently on vacation, so pull requests most likely won't be merged for a week or so still.

ZackingIt commented:
Happy new year's eve any chance this can get a review?  Click / signal detection while paused is pretty foundational, the only available work-arounds are pretty janky.

RandomShaper committed:
Implement pause-aware picking

This adds a new project setting (`physics/common/enable_pause_aware_picking`). It's disabled by default.

When enabled, it changes the way 2D & 3D physics picking behaves in relation to pause:
- When pause is set, every collision object that is hovered or captured (3D only) is released from that condition, getting the relevant mouse-exit callback., unless its pause mode makes it immune from pause.
- During the pause. picking only considers collision objects immune from pause, sending input events and enter/exit callbacks to them as expected.
- When pause is left, nothing happens. This is a big difference with the classic behavior, which at this point would process all the input events that have been queued against the current state of the 2D/3D world (in other words, checking them against the current position of the objects instead of those at the time of the events).

pouleyKetchoupp commented on scene/main/viewport.cpp:
```diff
@@ -418,288 +418,302 @@ void Viewport::_notification(int p_what) {
+		return;
+	if (!physics_object_picking)
+		return;
+	if (to_screen_rect == Rect2() || Input::get_singleton()->get_mouse_mode() != Input::MOUSE_MODE_CAPTURED)
```
This check seems inverted (see line 421 from the old code).

```suggestion
	if (to_screen_rect != Rect2() && Input::get_singleton()->get_mouse_mode() == Input::MOUSE_MODE_CAPTURED)
```

RandomShaper replied:
You're right. I actually added the right condition in the PR for _master_. To be honest, I knew I was going to mess the condition up somehow when rewriting it. 😅 I'll fix it in the next push.

pouleyKetchoupp commented on editor/project_manager.cpp:
```diff
@@ -499,6 +499,7 @@ class ProjectDialog : public ConfirmationDialog {
 					initial_settings["application/config/name"] = project_name->get_text();
 					initial_settings["application/config/icon"] = "res://icon.png";
 					initial_settings["rendering/environment/default_environment"] = "res://default_env.tres";
+					initial_settings["physics/common/enable_pause_aware_picking"] = true;
```
Shouldn't it be disabled on new projects if we decide to default to the old behavior on 3.2? Otherwise people might have compatibility breakage when they use an existing script on a different project.

RandomShaper replied:
My idea is to make it the new default so new projects benefit from the new behavior with zero setup.
However, you have a point in the cross-project compatibility rationale.
I don't have a strong opinion, but I think the benefit-to-risk ratio is higher with the new default.

pouleyKetchoupp replied:
Given that the old behavior is mostly buggy, on second thought I would agree it's best to introduce the new behavior by default on new projects.

pouleyKetchoupp reviewed (changes requested):
Nice work! This is very useful. I've tested with pause-immuned objects and pausable objects, and it was acting as expected (after the little fix from my review comment).

Also, before this is merged it would be nice to properly document this project option so users know what to expect.

RandomShaper commented:
Pushed with the fix for the condition and the new setting documented.

akien-mga commented:
Thanks!
